File: src/main/java/net/ndrei/teslacorelib/inventory/SidedFluidHandler.java
package net.ndrei.teslacorelib.inventory;

import com.google.common.collect.Lists;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.fluids.IFluidTank;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.ndrei.teslacorelib.capabilities.inventory.ISidedItemHandlerConfig;
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer;
import net.ndrei.teslacorelib.gui.FluidTankPiece;
import net.ndrei.teslacorelib.gui.IGuiContainerPiece;
import net.ndrei.teslacorelib.gui.IGuiContainerPiecesProvider;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.List;

/**
 * Created by CF on 2016-12-29.
 */
public class SidedFluidHandler extends FluidStorage implements ICapabilityProvider, IGuiContainerPiecesProvider {
    private ISidedItemHandlerConfig sidedConfig;

    public SidedFluidHandler(ISidedItemHandlerConfig sidedConfig) {
        this.sidedConfig = sidedConfig;
    }

    private List<IFluidTank> getTanksForSide(EnumFacing facing) {
        List<IFluidTank> list = Lists.newArrayList();

        for(IFluidTank tank: this.getTanks()) {
            if (tank instanceof ColoredFluidHandler) {
                EnumDyeColor color = ((ColoredFluidHandler)tank).getColor();
                if ((color != null) && ((facing == null) || this.sidedConfig.isSideSet(color, facing))) {
                    list.add(tank);
                }
            }
        }

        return list;
    }

    @Override
    public boolean hasCapability(@Nonnull Capability<?> capability, @Nullable EnumFacing facing) {
        if (capability == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) {
            List<IFluidTank> tanks = this.getTanksForSide(facing);
            return (tanks.size() > 0);
        }

        return false;
    }

    @Nullable
    @Override
    public <T> T getCapability(@Nonnull Capability<T> capability, @Nullable EnumFacing facing) {
        if (capability == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) {
            List<IFluidTank> tanks = this.getTanksForSide(facing);
            FluidStorage tempStorage = new FluidStorage();

            for(IFluidTank tank: tanks) {
                tempStorage.addTank(tank);
            }

            return (T)tempStorage;
        }

        return null;
    }

    @Override
    public List<IGuiContainerPiece> getGuiContainerPieces(BasicTeslaGuiContainer container) {
        List<IGuiContainerPiece> list = Lists.newArrayList();

        for(IFluidTank tank : this.getTanks()) {
            if (tank instanceof ColoredFluidHandler) {
                BoundingRectangle box = ((ColoredFluidHandler) tank).getBoundingBox();
                if (box != null) {
                    list.add(new FluidTankPiece(tank, box.getLeft(), box.getTop()));
                }
            }
        }

        return list;
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/inventory/SidedFluidHandler.kt
package net.ndrei.teslacorelib.inventory

import com.google.common.collect.Lists
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.fluids.IFluidTank
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.ndrei.teslacorelib.capabilities.inventory.ISidedItemHandlerConfig
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer
import net.ndrei.teslacorelib.gui.FluidTankPiece
import net.ndrei.teslacorelib.gui.IGuiContainerPiece
import net.ndrei.teslacorelib.gui.IGuiContainerPiecesProvider

/**
 * Created by CF on 2017-06-28.
 */
class SidedFluidHandler(private val sidedConfig: ISidedItemHandlerConfig)
    : FluidStorage(), ICapabilityProvider, IGuiContainerPiecesProvider, Iterable<IFluidTank> {

    private fun getTanksForSide(facing: EnumFacing?): List<IFluidTank> {
        val list = Lists.newArrayList<IFluidTank>()

        for (tank in this.tanks) {
            if (tank is ColoredFluidHandler) {
                val color = tank.color
                if (facing == null || this.sidedConfig.isSideSet(color, facing)) {
                    list.add(tank)
                }
            }
        }

        return list
    }

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        if (capability === CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) {
            val tanks = this.getTanksForSide(facing)
            return tanks.size > 0
        }

        return false
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        if (capability === CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) {
            val tanks = this.getTanksForSide(facing)
            val tempStorage = FluidStorage()

            for (tank in tanks) {
                tempStorage.addTank(tank)
            }

            return tempStorage as T
        }

        return null
    }

    override fun getGuiContainerPieces(container: BasicTeslaGuiContainer<*>): MutableList<IGuiContainerPiece> {
        val list = Lists.newArrayList<IGuiContainerPiece>()

        for (tank in this.tanks) {
            if (tank is ColoredFluidHandler) {
                val box = tank.boundingBox
                if (box != null) {
                    list.add(FluidTankPiece(tank, box.left, box.top))
                }
            }
        }

        return list
    }

    override fun iterator() = this.tanks.iterator()
}
